Generate IndexedDB object serializers

achristensen07 opened the pull request:
#### d4d3c5e42fefdd31a508016a05e758c4067e4915
<pre>
Generate IndexedDB object serializers
<a href="https://bugs.webkit.org/show_bug.cgi?id=244691">https://bugs.webkit.org/show_bug.cgi?id=244691</a>

Reviewed by Sihui Liu and Brady Eidson.

* Source/WebCore/Modules/indexeddb/IDBDatabaseIdentifier.cpp:
(WebCore::IDBDatabaseIdentifier::IDBDatabaseIdentifier):
* Source/WebCore/Modules/indexeddb/IDBDatabaseIdentifier.h:
(WebCore::IDBDatabaseIdentifier::encode const): Deleted.
(WebCore::IDBDatabaseIdentifier::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBFactory.cpp:
(WebCore::IDBFactory::openInternal):
(WebCore::IDBFactory::deleteDatabase):
* Source/WebCore/Modules/indexeddb/IDBGetAllResult.h:
(WebCore::IDBGetAllResult::IDBGetAllResult):
(WebCore::IDBGetAllResult::encode const): Deleted.
(WebCore::IDBGetAllResult::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBGetResult.h:
(WebCore::IDBGetResult::IDBGetResult):
(WebCore::IDBGetResult::encode const): Deleted.
(WebCore::IDBGetResult::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKeyRangeData.cpp:
(WebCore::IDBKeyRangeData::IDBKeyRangeData):
* Source/WebCore/Modules/indexeddb/IDBKeyRangeData.h:
(WebCore::IDBKeyRangeData::IDBKeyRangeData):
(WebCore::IDBKeyRangeData::encode const): Deleted.
(WebCore::IDBKeyRangeData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBValue.h:
(WebCore::IDBValue::encode const): Deleted.
(WebCore::IDBValue::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBCursorInfo.h:
(WebCore::IDBCursorInfo::transactionIdentifier const):
(WebCore::IDBCursorInfo::encode const): Deleted.
(WebCore::IDBCursorInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBCursorRecord.h:
(WebCore::IDBCursorRecord::isolatedCopy const):
(WebCore::IDBCursorRecord::encode const): Deleted.
(WebCore::IDBCursorRecord::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.cpp:
(WebCore::IDBDatabaseInfo::IDBDatabaseInfo):
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.h:
(WebCore::IDBDatabaseInfo::IDBDatabaseInfo):
(WebCore::IDBDatabaseInfo::encode const): Deleted.
(WebCore::IDBDatabaseInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseNameAndVersion.h:
(WebCore::IDBDatabaseNameAndVersion::encode const): Deleted.
(WebCore::IDBDatabaseNameAndVersion::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBError.h:
(WebCore::IDBError::messageForSerialization const):
(WebCore::IDBError::encode const): Deleted.
(WebCore::IDBError::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBGetAllRecordsData.h:
(WebCore::IDBGetAllRecordsData::encode const): Deleted.
(WebCore::IDBGetAllRecordsData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBGetRecordData.h:
(WebCore::IDBGetRecordData::encode const): Deleted.
(WebCore::IDBGetRecordData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBIndexInfo.h:
(WebCore::IDBIndexInfo::encode const): Deleted.
(WebCore::IDBIndexInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBIterateCursorData.h:
(WebCore::IDBIterateCursorData::encode const): Deleted.
(WebCore::IDBIterateCursorData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBObjectStoreInfo.cpp:
(WebCore::IDBObjectStoreInfo::IDBObjectStoreInfo):
* Source/WebCore/Modules/indexeddb/shared/IDBObjectStoreInfo.h:
(WebCore::IDBObjectStoreInfo::IDBObjectStoreInfo):
(WebCore::IDBObjectStoreInfo::encode const): Deleted.
(WebCore::IDBObjectStoreInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.h:
(WebCore::IDBRequestData::IDBRequestData):
(WebCore::IDBRequestData::encode const): Deleted.
(WebCore::IDBRequestData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
(WebCore::IDBResourceIdentifier::encode const): Deleted.
(WebCore::IDBResourceIdentifier::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResultData.h:
(WebCore::IDBResultData::encode const): Deleted.
(WebCore::IDBResultData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBTransactionInfo.cpp:
* Source/WebCore/Modules/indexeddb/shared/IDBTransactionInfo.h:
(WebCore::IDBTransactionInfo::IDBTransactionInfo):
(WebCore::IDBTransactionInfo::encode const): Deleted.
(WebCore::IDBTransactionInfo::decode): Deleted.
* Source/WebCore/bindings/js/JSIDBRequestCustom.cpp:
(WebCore::JSIDBRequest::result const):
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(MemberVariable.unique_ptr_type):
(generate_cpp):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::Subnamespace::StructName&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::encode):
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::ReturnRefClass&gt;::encode):
(IPC::ArgumentCoder&lt;Namespace::ReturnRefClass&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::EmptyConstructorStruct&gt;::encode):
(IPC::ArgumentCoder&lt;Namespace::EmptyConstructorStruct&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/254196@main">https://commits.webkit.org/254196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c32d59837da315a16ea632b7db6ec4960ead0421

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88370 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32815 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19208 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97557 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/153030 "Hash c32d5983 for PR 3934 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92336 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31307 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26960 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80569 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92215 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93978 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24904 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75218 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24883 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/91967 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79805 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67847 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28919 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14929 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32101 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/37844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1200 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31054 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34042 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/indexeddb/IDBKeyRangeData.h:
```diff
@@ -70,60 +79,17 @@ struct IDBKeyRangeData {
 
-    template<class Encoder> void encode(Encoder&) const;
-    template<class Decoder> static WARN_UNUSED_RETURN bool decode(Decoder&, IDBKeyRangeData&);
+    bool isNull;
```
IIRC @smfr made a change https://commits.webkit.org/236203@main to put the boolean variables together for memory optimization, so I am not sure we should change this.

achristensen07 replied:
ok.

szewai commented on Source/WebCore/Modules/indexeddb/shared/IDBRequestData.h:
```diff
@@ -71,12 +72,20 @@ class IDBRequestData {
 private:
+    friend struct IPC::ArgumentCoder<IDBRequestData, void>;
+    IDBRequestData(IDBConnectionIdentifier serverConnectionIdentifier, std::unique_ptr<IDBResourceIdentifier> requestIdentifier, std::unique_ptr<IDBResourceIdentifier> transactionIdentifier, std::unique_ptr<IDBResourceIdentifier> cursorIdentifier, uint64_t objectStoreIdentifier, uint64_t indexIdentifier, IndexedDB::IndexRecordType indexRecordType, std::optional<IDBDatabaseIdentifier> databaseIden ...
+        : m_serverConnectionIdentifier(WTFMove(serverConnectionIdentifier))
+        , m_requestIdentifier(WTFMove(requestIdentifier))
+        , m_transactionIdentifier(WTFMove(transactionIdentifier))
+        , m_cursorIdentifier(WTFMove(cursorIdentifier))
+        , m_objectStoreIdentifier(WTFMove(objectStoreIdentifier))
+        , m_indexIdentifier(WTFMove(indexIdentifier))
+        , m_indexRecordType(WTFMove(indexRecordType))
+        , m_databaseIdentifier(WTFMove(databaseIdentifier))
+        , m_requestedVersion(WTFMove(requestedVersion))
+        , m_requestType(WTFMove(requestType)) { }
```
No need to WTFMove for uint64_t

achristensen07 replied:
No reason not to, also.

szewai commented on Source/WebCore/Modules/indexeddb/IDBKeyRangeData.h:
```diff
@@ -62,6 +62,15 @@ struct IDBKeyRangeData {
+        : isNull(isNull)
+        , lowerKey(WTFMove(lowerKey))
+        , upperKey(WTFMove(upperKey))
+        , lowerOpen(WTFMove(lowerOpen))
+        , upperOpen(WTFMove(upperOpen))
```
No need for WTFMove

szewai commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -82,3 +82,182 @@ header: <WebCore/FloatPoint3D.h>
+
+header: <WebCore/IDBDatabaseInfo.h>
+class WebCore::IDBDatabaseInfo {
+    String m_name
+    uint64_t m_version
+    uint64_t m_maxIndexID
+    uint64_t m_maxObjectStoreID
+    HashMap<uint64_t, WebCore::IDBObjectStoreInfo> m_objectStoreMap
```
Why do we use `m_name` instead of `name()` like the other classes?

achristensen07 replied:
IDBDatabaseInfo has everything public, so we don't need to add "friend struct ICP::ArgumentCoder< IDBDatabaseInfo, void>;" and this is how were can access the things that need serializing.

szewai commented on Source/WebCore/Modules/indexeddb/shared/IDBCursorInfo.h:
```diff
@@ -65,19 +66,17 @@ class IDBCursorInfo {
     String loggingString() const;
 #endif
 
+    WEBCORE_EXPORT IDBCursorInfo(const IDBResourceIdentifier&, const IDBResourceIdentifier&, uint64_t, uint64_t, const IDBKeyRangeData&, IndexedDB::CursorSource, IndexedDB::CursorDirection, IndexedDB::CursorType);
```
Should this be public? (Asking as the other newly added constructor is private)

achristensen07 replied:
Private would be better.

szewai commented on Source/WebCore/bindings/js/JSIDBRequestCustom.cpp:
```diff
@@ -76,7 +76,7 @@ JSC::JSValue JSIDBRequest::result(JSC::JSGlobalObject& lexicalGlobalObject) cons
             return result ? result.value() : jsNull();
         });
-    }, [&] (IDBGetAllResult getAllResult) {
+    }, [&] (const IDBGetAllResult& getAllResult) {
```
I think we should either also updating the other parameters above (e.g. in `}, [&] (IDBGetResult getResult) {`), or leaving this change out.

achristensen07 replied:
I'll update all of them.  There are a lot of unnecessary copies here.

szewai commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -184,28 +199,50 @@ def generate_cpp(serialized_types, headers):
+                result.append('    return { ' + type.namespace_and_name() + '::create(')
+            else:
+                result.append('    return { ' + type.namespace_and_name() + ' {')
+            first_member = True
```
Where is this variable used?

achristensen07 replied:
Nothing.  It was used to determine whether to add a comma in an earlier version of this PR.  Removing.

szewai commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
```diff
@@ -118,7 +128,8 @@ std::optional<Namespace::OtherClass> ArgumentCoder<Namespace::OtherClass>::decod
 
-    return { {
+    return { Namespace::OtherClass {
+        WTFMove(*isNull),
```
No need to WTFMove

achristensen07 replied:
This is generated code.  I don't think it's a good idea to require the generator input to specify whether move is necessary when that would be error prone and have no benefit.

szewai commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
```diff
@@ -143,10 +157,86 @@ std::optional<Ref<Namespace::ReturnRefClass>> ArgumentCoder<Namespace::ReturnRef
+        return std::nullopt;
+
+    Namespace::EmptyConstructorStruct result;
+    result.m_int = WTFMove(*m_int);
+    result.m_double = WTFMove(*m_double);
```
No need to WTFMove

achristensen07 replied:
Ditto.

Commit:
Generate IndexedDB object serializers
https://bugs.webkit.org/show_bug.cgi?id=244691

Reviewed by Sihui Liu and Brady Eidson.

* Source/WebCore/Modules/indexeddb/IDBDatabaseIdentifier.cpp:
(WebCore::IDBDatabaseIdentifier::IDBDatabaseIdentifier):
* Source/WebCore/Modules/indexeddb/IDBDatabaseIdentifier.h:
(WebCore::IDBDatabaseIdentifier::encode const): Deleted.
(WebCore::IDBDatabaseIdentifier::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBFactory.cpp:
(WebCore::IDBFactory::openInternal):
(WebCore::IDBFactory::deleteDatabase):
* Source/WebCore/Modules/indexeddb/IDBGetAllResult.h:
(WebCore::IDBGetAllResult::IDBGetAllResult):
(WebCore::IDBGetAllResult::encode const): Deleted.
(WebCore::IDBGetAllResult::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBGetResult.h:
(WebCore::IDBGetResult::IDBGetResult):
(WebCore::IDBGetResult::encode const): Deleted.
(WebCore::IDBGetResult::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKeyRangeData.cpp:
(WebCore::IDBKeyRangeData::IDBKeyRangeData):
* Source/WebCore/Modules/indexeddb/IDBKeyRangeData.h:
(WebCore::IDBKeyRangeData::IDBKeyRangeData):
(WebCore::IDBKeyRangeData::encode const): Deleted.
(WebCore::IDBKeyRangeData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/IDBValue.h:
(WebCore::IDBValue::encode const): Deleted.
(WebCore::IDBValue::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBCursorInfo.h:
(WebCore::IDBCursorInfo::transactionIdentifier const):
(WebCore::IDBCursorInfo::encode const): Deleted.
(WebCore::IDBCursorInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBCursorRecord.h:
(WebCore::IDBCursorRecord::isolatedCopy const):
(WebCore::IDBCursorRecord::encode const): Deleted.
(WebCore::IDBCursorRecord::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.cpp:
(WebCore::IDBDatabaseInfo::IDBDatabaseInfo):
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.h:
(WebCore::IDBDatabaseInfo::IDBDatabaseInfo):
(WebCore::IDBDatabaseInfo::encode const): Deleted.
(WebCore::IDBDatabaseInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseNameAndVersion.h:
(WebCore::IDBDatabaseNameAndVersion::encode const): Deleted.
(WebCore::IDBDatabaseNameAndVersion::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBError.h:
(WebCore::IDBError::messageForSerialization const):
(WebCore::IDBError::encode const): Deleted.
(WebCore::IDBError::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBGetAllRecordsData.h:
(WebCore::IDBGetAllRecordsData::encode const): Deleted.
(WebCore::IDBGetAllRecordsData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBGetRecordData.h:
(WebCore::IDBGetRecordData::encode const): Deleted.
(WebCore::IDBGetRecordData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBIndexInfo.h:
(WebCore::IDBIndexInfo::encode const): Deleted.
(WebCore::IDBIndexInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBIterateCursorData.h:
(WebCore::IDBIterateCursorData::encode const): Deleted.
(WebCore::IDBIterateCursorData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBObjectStoreInfo.cpp:
(WebCore::IDBObjectStoreInfo::IDBObjectStoreInfo):
* Source/WebCore/Modules/indexeddb/shared/IDBObjectStoreInfo.h:
(WebCore::IDBObjectStoreInfo::IDBObjectStoreInfo):
(WebCore::IDBObjectStoreInfo::encode const): Deleted.
(WebCore::IDBObjectStoreInfo::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.cpp:
* Source/WebCore/Modules/indexeddb/shared/IDBRequestData.h:
(WebCore::IDBRequestData::IDBRequestData):
(WebCore::IDBRequestData::encode const): Deleted.
(WebCore::IDBRequestData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
(WebCore::IDBResourceIdentifier::encode const): Deleted.
(WebCore::IDBResourceIdentifier::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResultData.h:
(WebCore::IDBResultData::encode const): Deleted.
(WebCore::IDBResultData::decode): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBTransactionInfo.cpp:
* Source/WebCore/Modules/indexeddb/shared/IDBTransactionInfo.h:
(WebCore::IDBTransactionInfo::IDBTransactionInfo):
(WebCore::IDBTransactionInfo::encode const): Deleted.
(WebCore::IDBTransactionInfo::decode): Deleted.
* Source/WebCore/bindings/js/JSIDBRequestCustom.cpp:
(WebCore::JSIDBRequest::result const):
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(MemberVariable.unique_ptr_type):
(generate_cpp):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::Subnamespace::StructName>::decode):
(IPC::ArgumentCoder<Namespace::OtherClass>::encode):
(IPC::ArgumentCoder<Namespace::OtherClass>::decode):
(IPC::ArgumentCoder<Namespace::ReturnRefClass>::encode):
(IPC::ArgumentCoder<Namespace::ReturnRefClass>::decode):
(IPC::ArgumentCoder<Namespace::EmptyConstructorStruct>::encode):
(IPC::ArgumentCoder<Namespace::EmptyConstructorStruct>::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/254196@main
